doc/source/rllib/metrics-logger.rst

```diff
@@ -0,0 +1,431 @@
+
+.. _rllib-metric-logger-docs:
+
+MetricsLogger API
+==================
+
+.. include:: /_includes/rllib/new_api_stack.rst
+
+The RLlib team has designed the :py:class:`~ray.rllib.utils.metrics_logger.MetricsLogger` API
+to unify and make accessible the logging and processing of stats and metrics during your
+reinforcement learning (RL) experiments. RLlib's :py:class:`~ray.rllib.algorithms.algorithm.Algorithm`
+class and all its sub-components each have one :py:class:`~ray.rllib.utils.metrics_logger.MetricsLogger`
+instance managing metrics and statistics for this component. When a subcomponent reports back to its
+parent component, it "reduces" the logged results and sends them upstream.
+
+The RLlib team encourages you to use this API in all your custom code, like in
+:py:class:`~ray.rllib.env.env_runner.EnvRunner`-based :ref:`callbacks <rllib-callback-docs>`,
+in `custom loss functions <https://github.com/ray-project/ray/blob/master/rllib/examples/metrics/custom_metrics_in_loss_function.py>`__, or in custom `training_step() <https://github.com/ray-project/ray/blob/master/rllib/examples/metrics/custom_metrics_in_algorithm_training_step.py>`__
+implementations.
+
+.. figure:: images/metrics_logger_overview.svg
+    :width: 500
+    :align: left
+
+    **RLlib's MetricsLogger system**: Every subcomponent of an RLlib :py:class:`~ray.rllib.algorithms.algorithm.Algorithm` has-a
+    :py:class:`~ray.rllib.utils.metrics_logger.MetricsLogger` instance
+    and uses it to locally log values. When a component completes a distinct task,
+    for example, an :py:class:`~ray.rllib.env.env_runner.EnvRunner` finishing a sampling request, the local metrics of the subcomponent
+    (``EnvRunner``) are "reduced", and sent upstream to the containing parent component (``Algorithm``).
+    The parent component merges the received results into its own :py:class:`~ray.rllib.utils.metrics_logger.MetricsLogger` and,
+    at the end of its own task cycle, "reduces" as well for final reporting to the user or to Ray Tune.
+
+
+.. note::
+    So far, RLlib components owning a :py:class:`~ray.rllib.utils.metrics_logger.MetricsLogger`
+    instance are :py:class:`~ray.rllib.algorithms.algorithm.Algorithm`, :py:class:`~ray.rllib.env.env_runner.EnvRunner`,
+    :py:class:`~ray.rllib.core.learner.learner.Learner`, and all :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` classes.
+    The Ray team is considering expanding access to this API on other components as well.
+
+
+Features of MetricsLogger
+-------------------------
+
+The :py:class:`~ray.rllib.utils.metrics_logger.MetricsLogger` API offers the following features:
+
+- Users can log scalar values over time, such as losses or rewards.
+- Thereby, users can configure different reduction types, in particular ``mean``, ``min``, ``max``, ``sum``, or ``None``.
+- Users can also specify sliding windows, over which these reductions take place, for example ``window=100`` to average over the
```

Comment:
Maybe say something about `clear_on_reduce` and how aggregates are held in the history instead of accumulating large lists.

Reply:
done